src/python/pants/engine/target.py

```diff
@@ -36,6 +36,7 @@
 @dataclass(eq=False)  # type: ignore[misc]   # https://github.com/python/mypy/issues/5374
 class Field(ABC):
     alias: ClassVar[str]
+    default: ClassVar[Any]
```

Comment:
Even if we can't parameterize `Field` in higher layers for techinical resaons, it seems to make sense to do so here and let `default` / `hydrate`, etc take the real type they work with instead of `Any`.

Reply:
Subclasses are expected to override the type hints with what they use:

https://github.com/pantsbuild/pants/blob/23c4bf6fa06aa0330fdd2b0c893fcbecacddca2c/src/python/pants/engine/target.py#L610-L618

We expect the vast majority of fields to subclass these helper types like `BoolField` and `StringField`. So far, when writing bindings, they've been able to capture every Field I've encountered outside of `provides`.

--

Parametrization doesn't quite work because this is describing an interface with an outside world, not between Python code. We're saying "I expect the user will use an `int` in their BUILD file", but we have no guarantee and the only way we can figure out the type of what the actually passed is via `type()` and `isinstance()`. In contrast, MyPy needs to be able to statically determine the type of the generic.